ocaml/tests/test_vm_helpers.ml

```diff
@@ -573,6 +609,263 @@ let test_get_group_key_vgpu_and_netsriov () =
+    ; resident_on= [("vm1", "h1"); ("vm2", "h2"); ("vm3", "h3")]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"; "h2"; "h3"]]
+    }
+  ; {
+      description= "3 running VMs (h1(0) h2(1) h3(2))"
+    ; state= "running"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h2"); ("vm2", "h3"); ("vm3", "h3")]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"]; ["h2"]; ["h3"]]
+    }
+  ; {
+      description= "3 running VMs (h1(0) h2(0) h3(3))"
+    ; state= "running"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h3"); ("vm2", "h3"); ("vm3", "h3")]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"; "h2"]; ["h3"]]
+    }
+  ; {
+      description= "3 starting VMs (h1(1) h2(1) h3(1))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h1"); ("vm2", "h2"); ("vm3", "h3")]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"; "h2"; "h3"]]
+    }
+  ; {
+      description= "3 starting VMs (h1(0) h2(1) h3(2))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h2"); ("vm2", "h3"); ("vm3", "h3")]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"]; ["h2"]; ["h3"]]
+    }
+  ; {
+      description= "3 starting VMs (h1(0) h2(0) h3(3))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h3"); ("vm2", "h3"); ("vm3", "h3")]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"; "h2"]; ["h3"]]
+    }
+  ; {
+      description= "3 running VMs with affinity host"
+    ; state= "running"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h1"); ("vm2", "h2"); ("vm3", "h3")]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= Some "h3"
+    ; expected= [["h3"]; ["h1"; "h2"]]
+    }
+  ; {
+      description= "6 running VMs (h1(2) h2(2) h3(2))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"; "vm4"; "vm5"; "vm6"]
+    ; resident_on=
+        [
+          ("vm1", "h1")
+        ; ("vm2", "h1")
+        ; ("vm3", "h2")
+        ; ("vm4", "h2")
+        ; ("vm5", "h3")
+        ; ("vm6", "h3")
+        ]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"; "h2"; "h3"]]
+    }
+  ; {
+      description= "6 running VMs (h1(1) h2(2) h3(3))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"; "vm4"; "vm5"; "vm6"]
+    ; resident_on=
+        [
+          ("vm1", "h1")
+        ; ("vm2", "h2")
+        ; ("vm3", "h2")
+        ; ("vm4", "h3")
+        ; ("vm5", "h3")
+        ; ("vm6", "h3")
+        ]
+    ; placement_policy= "anti_affinity"
+    ; affinity_host= None
+    ; expected= [["h1"]; ["h2"]; ["h3"]]
+    }
+  ]
+
+let make_other_vms_for_aa_test ~__context ~host1 ~host2 ~host3 =
```

Comment:
This is hard coded for each test case now, it would be better to make the none-anti-affinity VMs and VMs in other AA group configurable.